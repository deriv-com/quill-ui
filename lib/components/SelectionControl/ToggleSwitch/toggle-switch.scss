@import "@quill/color.scss";
@import "@quill/color.opacity.scss";
@import "@quill/others.scss";
@import "@quill/border.scss";
@import "@quill/motion.scss";
@import "@quill/size.scss";
@import "@quill/spacing.scss";

.toggle-switch {
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
    user-select: none;
    width: var(--core-size-2400);
    height: var(--core-size-1400);
    border-width: var(--core-borderRadius-100);
    &__disabled .toggle-switch__slider {
        pointer-events: none;
        background: var(--core-color-opacity-black-200);
    }
    &__slider {
        display: flex;
        align-items: center;
        justify-content: flex-start;
        width: var(--core-size-2200);
        height: var(--size-generic-sm);
        background-color: var(--core-color-opacity-black-200);
        border-radius: var(--semantic-borderRadius-lg);
        transition: background-color var(--core-motion-duration-300) ease;
        cursor: pointer;
        position: relative;
        &:hover:not(.disabled) {
            background-color: var(--core-color-opacity-black-300);
        }

        &:active:not(.disabled) {
            background-color: var(--core-color-opacity-black-400);
        }

        &.enabled {
            background-color: var(--core-color-opacity-green-600);
            &:hover {
                background-color: var(--core-color-opacity-green-800);
            }

            &:active {
                background-color: var(--core-color-solid-green-900);
            }
        }
    }
    &__knob {
        position: absolute;
        width: var(--core-size-1000);
        height: var(--core-size-1000);
        border-radius: var(--core-borderRadius-2500);
        background-color: var(--core-color-solid-slate-50);
        transition: left var(--core-motion-duration-300) ease;
        top: 50%;
        transform: translateY(-50%);
        left: var(--core-spacing-100);

        &.enabled {
            left: calc(100% - 22px);
        }

        &.disabled {
            background-color: var(--core-color-opacity-white-400);
        }
    }
}
