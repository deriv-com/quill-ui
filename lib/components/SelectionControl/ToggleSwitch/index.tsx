import React, { useState } from "react";
import "./toggle-switch.scss"; // Ensure your SCSS file path is correct
import clsx from "clsx";

export interface SwitchToggleProps {
    disabled?: boolean;
    defaultChecked?: boolean;
    onChange?: (isEnabled: boolean) => void;
    className?: string;
    id?: string;
    onFocus?: () => void;
    onBlur?: () => void;
}

const ToggleSwitch = ({
    disabled = false,
    onChange,
    className,
    defaultChecked,
    onFocus,
    onBlur,
    id,
}: SwitchToggleProps) => {
    const [isEnabled, setIsEnabled] = useState(defaultChecked || false);

    const toggleSwitch = () => {
        if (!disabled) {
            const toggleEnable = !isEnabled;
            setIsEnabled(toggleEnable);
            onChange?.(toggleEnable);
        }
    };

    const handleKeyDown = (event: React.KeyboardEvent) => {
        if (event.key === "Enter" || event.key === " ") {
            event.preventDefault();
            toggleSwitch();
        }
    };

    return (
        <div
            className={clsx("toggle-switch", className, { disabled })}
            tabIndex={disabled ? -1 : 0}
            onClick={toggleSwitch}
            onKeyDown={handleKeyDown}
            role="button"
            aria-pressed={isEnabled}
            onFocus={onFocus}
            id={id}
            onBlur={onBlur}
            aria-disabled={disabled}
        >
            <div
                className={clsx("toggle-switch__slider", {
                    enabled: isEnabled,
                })}
            >
                <div className={clsx("toggle-switch__knob", { enabled: isEnabled, disabled })}></div>
            </div>
        </div>
    );
};

export default ToggleSwitch;
