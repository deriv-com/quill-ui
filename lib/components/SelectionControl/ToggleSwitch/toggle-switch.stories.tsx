import { StoryObj, Meta } from "@storybook/react";
import ToggleSwitch from ".";
import "./toggle-switch.scss";
import { fn } from "@storybook/test";

const meta = {
    title: "Components/Selection control/Toggle Switch",
    component: ToggleSwitch,
    parameters: { layout: "centered" },
    tags: ["autodocs"],
    args: {
        defaultChecked: false,
        disabled: false,
        onChange: fn(),
        onFocus: fn(),
        onBlur: fn(),
        className: "",
    },
    argTypes: {
        defaultChecked: {
            table: { type: { summary: "boolean | undefined" } },
            description: "Flag for setting initial state",
            control: { type: "boolean" },
        },
        checked: {
            table: { type: { summary: "boolean | undefined" } },
            description: "Setting the switch state from outside the component and manage it outside, It enforces whichever state is coming",
            control: { type: "boolean" },
        },
        disabled: {
            table: { type: { summary: "boolean | undefined" } },
            description: "Flag for setting component's disabled state",
            control: { type: "boolean" },
        },
        className: {
            table: { type: { summary: "string | undefined" } },
            description: "ClassName for the switch's wrapper",
            control: { type: "text" },
        },
        id: {
            table: { type: { summary: "string | undefined" } },
            description:
                "Input's ID attribute value, for associating the label",
            control: { type: "text" },
        },
        onChange: {
            table: { type: { summary: "function | undefined" } },
            description:
                "Callback function called when the switch's value changes",
            control: { type: null },
        },
        onFocus: {
            table: { type: { summary: "function | undefined" } },
            description: "Callback function called when the switch gains focus",
            control: { type: null },
        },
        onBlur: {
            table: { type: { summary: "function | undefined" } },
            description: "Callback function called when the switch loses focus",
            control: { type: null },
        },
    },
} satisfies Meta<typeof ToggleSwitch>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        disabled: false,
        defaultChecked: false,
    },
};

export const DefaultOn: Story = {
    args: {
        disabled: false,
        defaultChecked: true,
    },
    name:'Checked (DefaultChecked is true)'
};

export const Disabled: Story = {
    args: {
        disabled: true,
        defaultChecked: false,
    },
};

export const DisabledOn: Story = {
    args: {
        disabled: true,
        defaultChecked: true,
    },
};

export const Checked_StayAlwaysChecked: Story = {
    args: {
        checked:true,
        disabled: false,
    },
    name:'Checked (Stay Always Checked)'
};