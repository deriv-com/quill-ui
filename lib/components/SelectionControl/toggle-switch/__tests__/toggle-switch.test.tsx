import React from "react";
import { act, fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import ToggleSwitch from "../index";

describe("ToggleSwitch", () => {
    it("should toggle from off to on", () => {
        const onChange = jest.fn();
        const { getByRole } = render(<ToggleSwitch onChange={onChange} />);
        const switchEl = getByRole("button");
        act(() => {
            fireEvent.click(switchEl);
        });
        expect(onChange).toHaveBeenCalledWith(true);
    });

    it("should toggle from on to off when defaultChecked", () => {
        const onChange = jest.fn();
        const { getByRole } = render(
            <ToggleSwitch defaultChecked onChange={onChange} />,
        );
        const switchEl = getByRole("button");
        act(() => {
            fireEvent.click(switchEl);
        });
        expect(onChange).toHaveBeenCalledWith(false);
    });

    it("should not toggle when disabled", () => {
        const onChange = jest.fn();
        const { getByRole } = render(
            <ToggleSwitch disabled onChange={onChange} />,
        );
        const switchEl = getByRole("button");
        act(() => {
            fireEvent.click(switchEl);
        });
        expect(onChange).not.toHaveBeenCalled();
    });

    it("should not toggle when checked is true", () => {
        const onChange = jest.fn();
        const { getByRole } = render(
            <ToggleSwitch checked onChange={onChange} />,
        );
        const switchEl = getByRole("button");
        act(() => {
            fireEvent.click(switchEl);
        });
        expect(onChange).toHaveBeenCalledWith(false);
    });

    it("should be focusable and trigger onFocus and onBlur", () => {
        const onFocus = jest.fn();
        const onBlur = jest.fn();
        const { getByRole } = render(
            <ToggleSwitch onFocus={onFocus} onBlur={onBlur} />,
        );
        const switchEl = getByRole("button");
        act(() => {
            fireEvent.focus(switchEl);
        });
        expect(onFocus).toHaveBeenCalled();
        act(() => {
            fireEvent.blur(switchEl);
        });
        expect(onBlur).toHaveBeenCalled();
    });

    it("renders correctly with default props", () => {
        const { asFragment } = render(<ToggleSwitch onChange={() => {}} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("toggles from off to on and matches snapshot", () => {
        const onChange = jest.fn();
        const { getByRole, asFragment } = render(
            <ToggleSwitch onChange={onChange} />,
        );
        act(() => {
            fireEvent.click(getByRole("button"));
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly when disabled", () => {
        const { asFragment } = render(
            <ToggleSwitch disabled onChange={() => {}} />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly when defaultChecked", () => {
        const { asFragment } = render(
            <ToggleSwitch defaultChecked onChange={() => {}} />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot when focused", () => {
        const onFocus = jest.fn();
        const { getByRole, asFragment } = render(
            <ToggleSwitch onChange={() => {}} onFocus={onFocus} />,
        );
        const toggleElement = getByRole("button");
        act(() => {
            fireEvent.focus(toggleElement);
        });
        expect(onFocus).toHaveBeenCalled();
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot when blurred", () => {
        const onBlur = jest.fn();
        const { getByRole, asFragment } = render(
            <ToggleSwitch onChange={() => {}} onBlur={onBlur} />,
        );
        const toggleElement = getByRole("button");
        act(() => {
            fireEvent.focus(toggleElement);
            fireEvent.blur(toggleElement);
        });
        expect(onBlur).toHaveBeenCalled();
        expect(asFragment()).toMatchSnapshot();
    });
});
