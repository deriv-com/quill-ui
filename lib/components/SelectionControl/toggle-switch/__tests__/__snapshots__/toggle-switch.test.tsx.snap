// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch matches snapshot when blurred 1`] = `
<DocumentFragment>
  <button
    aria-pressed="false"
    class="toggle-switch"
    data-testid="toggle-switch"
  >
    <div
      class="toggle-switch__slider"
    >
      <div
        class="toggle-switch__knob"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`ToggleSwitch matches snapshot when focused 1`] = `
<DocumentFragment>
  <button
    aria-pressed="false"
    class="toggle-switch"
    data-testid="toggle-switch"
  >
    <div
      class="toggle-switch__slider"
    >
      <div
        class="toggle-switch__knob"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`ToggleSwitch renders correctly when defaultChecked 1`] = `
<DocumentFragment>
  <button
    aria-pressed="true"
    class="toggle-switch"
    data-testid="toggle-switch"
  >
    <div
      class="toggle-switch__slider enabled"
    >
      <div
        class="toggle-switch__knob enabled"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`ToggleSwitch renders correctly when disabled 1`] = `
<DocumentFragment>
  <button
    aria-pressed="false"
    class="toggle-switch"
    data-testid="toggle-switch"
    disabled=""
  >
    <div
      class="toggle-switch__slider"
    >
      <div
        class="toggle-switch__knob"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`ToggleSwitch renders correctly with default props 1`] = `
<DocumentFragment>
  <button
    aria-pressed="false"
    class="toggle-switch"
    data-testid="toggle-switch"
  >
    <div
      class="toggle-switch__slider"
    >
      <div
        class="toggle-switch__knob"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`ToggleSwitch toggles from off to on and matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-pressed="true"
    class="toggle-switch"
    data-testid="toggle-switch"
  >
    <div
      class="toggle-switch__slider enabled"
    >
      <div
        class="toggle-switch__knob enabled"
      />
    </div>
  </button>
</DocumentFragment>
`;
